File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        max_len = 0
        left = 0
        freq = [0] * 26
        max_freq = 0
        for right in range(len(s)):
            freq[ord(s[right]) - ord('A')] += 1
            max_freq = max(max_freq, freq[ord(s[right]) - ord('A')])

            while (right - left + 1) - max_freq > k:
                freq[ord(s[left]) - ord('A')] -= 1
                left += 1
            
            max_len = max(max_len, right - left + 1)
        return max_len

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int left = 0;
        int max_len = 0;
        int[] freq = new int[26];
        int max_freq = 0;

        for(int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            freq[ch - 'A'] += 1;
            max_freq = Math.max(max_freq, freq[ch - 'A']);
        
            while((right - left + 1) - max_freq > k) {
                ch = s.charAt(left);
                freq[ch - 'A'] -= 1;
                left += 1;
            }
            max_len = Math.max(max_len, (right - left + 1));
        }
        return max_len;
    }
}